File: Daily/100.same-tree.js
/*
 * @lc app=leetcode id=100 lang=javascript
 *
 * [100] Same Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode | null} p
 * @param {TreeNode | null} q
 * @return {boolean}
 */
var isSameTree = function (p, q) {
    if ((p === null && q !== null) || (p !== null && q === null)) return false;
    else if (p !== null && q !== null) {
        if (p.val !== q.val) {
            return false;
        } else {
            return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
        }
    } else {
        // p === null && q === null
        return true;
    }
};
// @lc code=end

File: prev-2024/100.same-tree.cpp
/*
 * @lc app=leetcode id=100 lang=cpp
 *
 * [100] Same Tree
 */

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;

  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};

// @lc code=start
class Solution {
 public:
  bool isSameTree(TreeNode* p, TreeNode* q) {
    if (p == nullptr && q != nullptr || p != nullptr && q == nullptr) {
      return false;
    }
    if (p == nullptr && q == nullptr) {
      return true;
    }
    return p->val == q->val && isSameTree(p->left, q->left) &&
           isSameTree(p->right, q->right);
  }
};
// @lc code=end
